### PR TITLE
[FLINK-29900][Connectors/DynamoDB] Implement Table API for DynamoDB sink

### PR DESCRIPTION
**Pull request by hlteoh37**
## What is the purpose of the change
- Implement Table API for DynamoDB sink

* Note that the support for Doubles is not great at the moment. The Java limitations for Double is more permissive than the DDB limitation for the number type. This means that if a user writes a record with any field of DOUBLE value larger than permitted by DynamoDB, the record will fail to write.

We will followup with [FLINK-30092](https://issues.apache.org/jira/browse/FLINK-30092) to improve the experience here.

## Does this pull request potentially affect one of the following parts:
* Dependencies (does it add or upgrade a dependency): yes, table API
* The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
* The serializers: yes
* The runtime per-record code paths (performance sensitive): yes
* Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no

## Documentation
* Does this pull request introduce a new feature? yes
* If yes, how is the feature documented?

### CODE REVIEW

File: flink-connector-dynamodb/pom.xml [hunk 1]
```diff
@@ -77,6 +77,22 @@ under the License.
...
+			<scope>provided</scope>
+			<optional>true</optional>
+		</dependency>
+		<dependency>
+            <groupId>org.apache.flink</groupId>
+            <artifactId>flink-table-runtime</artifactId>
+            <version>${flink.version}</version>
+            <scope>provided</scope>
+            <optional>true</optional>
+        </dependency>
```
Comment (dannycranmer) [inline flink-connector-dynamodb/pom.xml:94]: Formatting looks squiffy

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package org.apache.flink.connector.dynamodb.table;
+
+import org.apache.flink.configuration.ReadableConfig;
+
+import static org.apache.flink.connector.dynamodb.table.DynamoDbConnectorOptions.FAIL_ON_ERROR;
+import static org.apache.flink.connector.dynamodb.table.DynamoDbConnectorOptions.TABLE_NAME;
+
+/** DynamoDb specific configuration. */
+public class DynamoDbConfiguration {
```
Comment (dannycranmer) [inline flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbConfiguration.java:9]: Missing support annotation. Please make non-test classes have `@Internal/@PublicEvolving` annotation

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package org.apache.flink.connector.dynamodb.table;
```
Comment (dannycranmer) [inline flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbConfiguration.java:1]: Missing copyright, please check all other files too

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbConnectorOptions.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
+package org.apache.flink.connector.dynamodb.table;
+
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.configuration.ConfigOptions;
+
+/**
+ * Base options for the DynamoDb connector. Needs to be public so that the {@link
+ * org.apache.flink.table.api.TableDescriptor} can access it.
```
Comment (dannycranmer) [inline flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbConnectorOptions.java:8]: nit: add an import for TableDescriptor and simplify comment

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbConnectorOptions.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ */
+public class DynamoDbConnectorOptions {
+
+    private DynamoDbConnectorOptions() {
+        // private constructor to prevent initialization of static class
+    }
+
+    public static final ConfigOption<String> TABLE_NAME =
+            ConfigOptions.key("table")
```
Comment (dannycranmer) [inline flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbConnectorOptions.java:17]: Should we keep this consistent with the DataStream field? `destination-table-name` or rename both to `table-name`?

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbDynamicSink.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+    @Override
+    public String asSummaryString() {
+        // TODO: check when this is called/returned
+        return "DynamoDB";
+    }
+
+    @Override
+    public void applyStaticPartition(Map<String, String> partitions) {
+        this.overwriteByPartitionKeys.addAll(partitions.keySet());
```
Comment (dannycranmer) [inline flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbDynamicSink.java:107]: nit: remove `this`

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbDynamicSink.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+    @Override
+    public int hashCode() {
+        return Objects.hash(
+                super.hashCode(),
+                destinationTableName,
+                failOnError,
+                dynamoDbClientProperties,
+                physicalDataType,
+                overwriteByPartitionKeys);
+    }
```
Comment (dannycranmer) [inline flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbDynamicSink.java:142]: Why are these needed? Please review https://flink.apache.org/contributing/code-style-and-quality-java.html#equals--hashcode and remove if we can

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbDynamicSinkFactory.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+import org.apache.flink.shaded.guava30.com.google.common.collect.ImmutableSet;
+
+import java.util.HashSet;
+import java.util.Properties;
+import java.util.Set;
+
+import static org.apache.flink.connector.dynamodb.table.DynamoDbConnectorOptions.TABLE_NAME;
+
+/** */
```
Comment (dannycranmer) [inline flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbDynamicSinkFactory.java:17]: todo

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataElementConverter.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+import org.apache.flink.api.connector.sink2.SinkWriter;
+import org.apache.flink.connector.base.sink.writer.ElementConverter;
+import org.apache.flink.connector.dynamodb.sink.DynamoDbWriteRequest;
+import org.apache.flink.connector.dynamodb.sink.DynamoDbWriteRequestType;
+import org.apache.flink.table.api.TableException;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.types.DataType;
+
+/** Implementation of an {@link ElementConverter} specific for DynamoDb sink. */
```
Comment (dannycranmer) [inline flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataElementConverter.java:11]: Please elaborate here

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataElementConverter.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+                new RowDataToAttributeValueConverter(physicalDataType);
+    }
+
+    @Override
+    public DynamoDbWriteRequest apply(RowData element, SinkWriter.Context context) {
+        switch (element.getRowKind()) {
+            case INSERT:
+            case UPDATE_AFTER:
+                return getPutRequest(element);
+            case UPDATE_BEFORE:
```
Comment (dannycranmer) [inline flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataElementConverter.java:29]: Why does `UPDATE_BEFORE` map to delete?

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataElementConverter.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+                .setItem(rowDataToAttributeValueConverter.convertRowData(row))
+                .build();
+    }
+
+    private DynamoDbWriteRequest getDeleteRequest(RowData row) {
+        return DynamoDbWriteRequest.builder()
+                .setType(DynamoDbWriteRequestType.DELETE)
+                .setItem(rowDataToAttributeValueConverter.convertRowData(row))
+                .build();
+    }
```
Comment (dannycranmer) [inline flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataElementConverter.java:49]: nit: I like this, however you could create the builder in the previous method

```
@Override
    public DynamoDbWriteRequest apply(RowData element, SinkWriter.Context context) {
      var builder = DynamoDbWriteRequest.builder()
                .setItem(rowDataToAttributeValueConverter.convertRowData(row));
    
        switch (element.getRowKind()) {
            case INSERT:
            case UPDATE_AFTER:
                builder.setType(DynamoDbWriteRequestType.PUT);
            case UPDATE_BEFORE:
            case DELETE:
                builder.setType(DynamoDbWriteRequestType.DELETE);
            default:
                throw new TableException("Unsupported message kind: " + element.getRowKind());
        }

      return builder.build();
    }
```

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataToAttributeValueConverter.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+                                                (T)
+                                                        DataStructureConverters.getConverter(
+                                                                        field.getDataType())
+                                                                .toExternal(
+                                                                        fieldGetter.getFieldOrNull(
+                                                                                rowData)))
+                                .setter(((rowData, t) -> {})));
+    }
+
+    public Map<String, AttributeValue> convertRowData(RowData row) {
```
Comment (dannycranmer) [inline flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataToAttributeValueConverter.java:66]: nit: Public methods should typically go above private, for readability

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataToAttributeValueConverter.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+                                .getter(
+                                        rowData ->
+                                                (T)
+                                                        DataStructureConverters.getConverter(
+                                                                        field.getDataType())
+                                                                .toExternal(
+                                                                        fieldGetter.getFieldOrNull(
+                                                                                rowData)))
+                                .setter(((rowData, t) -> {})));
+    }
```
Comment (dannycranmer) [inline flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataToAttributeValueConverter.java:64]: Yikes, spotless added some spots here :/

**Review by dannycranmer — CHANGES_REQUESTED** [review]

File: flink-connector-dynamodb/pom.xml (hunk 1 above)
Comment (hlteoh37) [reply flink-connector-dynamodb/pom.xml]: Addressed. Good spot

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataElementConverter.java (hunk 11 above)
Comment (hlteoh37) [reply flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataElementConverter.java]: Ok, done!

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataToAttributeValueConverter.java (hunk 12 above)
Comment (hlteoh37) [reply flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataToAttributeValueConverter.java]: done

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataElementConverter.java (hunk 9 above)
Comment (hlteoh37) [reply flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataElementConverter.java]: done

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbDynamicSink.java (hunk 7 above)
Comment (hlteoh37) [reply flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbDynamicSink.java]: Yep, these were only used for tests. Removed and improved tests.

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataToAttributeValueConverter.java (hunk 13 above)
Comment (hlteoh37) [reply flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataToAttributeValueConverter.java]: :O Can't think of a nice way to simplify this...! It has 2 nested lambdas, probably why spotless went mad...

**Review by hlteoh37 — COMMENTED** [review]
Addressed comments!

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataToAttributeValueConverter.java [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.flink.table.data.RowData.createFieldGetter;
+
+/** Converts from Flink Table API internal type of {@link RowData} to {@link AttributeValue}. */
+@Internal
+public class RowDataToAttributeValueConverter implements Serializable {
```
Comment (dannycranmer) [inline flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataToAttributeValueConverter.java:43]: nit: You _could_ create this class lazily in `RowDataElementConverter` so this does not need to be serialised/serializable.

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataElementConverter.java [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+public class RowDataElementConverter implements ElementConverter<RowData, DynamoDbWriteRequest> {
+
+    private final DataType physicalDataType;
+    private final RowDataToAttributeValueConverter rowDataToAttributeValueConverter;
+
+    public RowDataElementConverter(DataType physicalDataType) {
+        this.physicalDataType = physicalDataType;
+        this.rowDataToAttributeValueConverter =
+                new RowDataToAttributeValueConverter(physicalDataType);
+    }
```
Comment (dannycranmer) [inline flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataElementConverter.java:45]: For `DynamoDbBeanElementConverter` I create the table schema here to fail fast. Currently you do not create the schema until consuming data, what happens if table invalid?  https://github.com/apache/flink-connector-aws/blob/main/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/sink/DynamoDbBeanElementConverter.java#L52

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/converter/ArrayAttributeConverterProvider.java [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+        } else if (enhancedType.equals(EnhancedType.of(LocalTime[].class))) {
+            return (AttributeConverter<T>)
+                    getArrayAttributeConverter(LocalTime.class, enhancedType);
+        } else if (enhancedType.equals(EnhancedType.of(LocalDateTime[].class))) {
+            return (AttributeConverter<T>)
+                    getArrayAttributeConverter(LocalDateTime.class, enhancedType);
+        } else if (enhancedType.equals(EnhancedType.of(Instant[].class))) {
+            return (AttributeConverter<T>) getArrayAttributeConverter(Instant.class, enhancedType);
+        }
+        return null;
```
Comment (dannycranmer) [inline flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/converter/ArrayAttributeConverterProvider.java:71]: What happens in this failure mode? Should we throw an exception here?

As discussed, you missed `Double`

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/converter/ArrayAttributeConverterProvider.java [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+
+/** Attribute converter provider for String Array. */
+@Internal
+public class ArrayAttributeConverterProvider implements AttributeConverterProvider {
+
+    private static final AttributeConverterProvider defaultAttributeConverterProvider =
+            AttributeConverterProvider.defaultProvider();
+
+    @Override
+    public <T> AttributeConverter<T> converterFor(EnhancedType<T> enhancedType) {
```
Comment (dannycranmer) [inline flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/converter/ArrayAttributeConverterProvider.java:41]: How do you propose we keep this method in sync between Flink types and `EnhancedType` https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/EnhancedType.java ?

**Review by dannycranmer — CHANGES_REQUESTED** [review]

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataToAttributeValueConverter.java (hunk 13 above)
Comment (dannycranmer) [reply flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataToAttributeValueConverter.java]: You _could_ extract the lambdas, but it is a nit

**Review by dannycranmer — CHANGES_REQUESTED** [review]

File: flink-sql-connector-dynamodb/pom.xml [hunk 18]
```diff
@@ -77,12 +77,16 @@ under the License.
...
+									<include>commons-codec:commons-codec</include>
 								</includes>
 							</artifactSet>
 							<relocations>
 								<relocation>
 									<pattern>software.amazon</pattern>
 									<shadedPattern>org.apache.flink.connector.dynamodb.shaded.software.amazon</shadedPattern>
+									<excludes>
+										<exclude>software.amazon.awssdk.enhanced.**</exclude>
+									</excludes>
```
Comment (dannycranmer) [inline flink-sql-connector-dynamodb/pom.xml:89]: Why are these not relocated?

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataToAttributeValueConverter.java (hunk 14 above)
Comment (hlteoh37) [reply flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataToAttributeValueConverter.java:43]: done

File: flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbConfiguration.java [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.configuration.ReadableConfig;
+
+import static org.apache.flink.connector.dynamodb.table.DynamoDbConnectorOptions.FAIL_ON_ERROR;
+import static org.apache.flink.connector.dynamodb.table.DynamoDbConnectorOptions.TABLE_NAME;
+
+/** DynamoDb specific configuration. */
+@PublicEvolving
```
[the remark this answers is not in the document]
Comment (hlteoh37) [reply flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/DynamoDbConfiguration.java:28]: done.

**Review by hlteoh37 — COMMENTED** [review]
addressed comments

File: flink-sql-connector-dynamodb/pom.xml (hunk 18 above)
Comment (hlteoh37) [reply flink-sql-connector-dynamodb/pom.xml:89]: We cannot shade these because EnhancedDynamoDB uses an additional interceptor in the AWS SDK chain. The ChainFactory resolves this a manner that the shade plugin doesn't handle properly.

https://github.com/aws/aws-sdk-java-v2/blob/master/core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/ClasspathInterceptorChainFactory.java#L42


```
Caused by: java.lang.ClassNotFoundException: software.amazon.awssdk.enhanced.dynamodb.internal.ApplyUserAgentInterceptor
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:476)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
	at org.apache.flink.util.FlinkUserCodeClassLoader.loadClassWithoutExceptionHandling(FlinkUserCodeClassLoader.java:68)
	at org.apache.flink.util.ChildFirstClassLoader.loadClassWithoutExceptionHandling(ChildFirstClassLoader.java:74)
	at org.apache.flink.util.FlinkUserCodeClassLoader.loadClass(FlinkUserCodeClassLoader.java:52)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:315)
	at org.apache.flink.streaming.connectors.dynamodb.shaded.software.amazon.awssdk.core.internal.util.ClassLoaderHelper.loadClass(ClassLoaderHelper.java:115)
	at org.apache.flink.streaming.connectors.dynamodb.shaded.software.amazon.awssdk.core.internal.util.ClassLoaderHelper.loadClass(ClassLoaderHelper.java:74)
	at org.apache.flink.streaming.connectors.dynamodb.shaded.software.amazon.awssdk.core.interceptor.ClasspathInterceptorChainFactory.createExecutionInterceptor(ClasspathInterceptorChainFactory.java:123)
	... 37 more

```

File: flink-connector-dynamodb/src/test/java/org/apache/flink/connector/dynamodb/table/DynamoDbDynamicSinkFactoryTest.java [hunk 20]
```diff
@@ -0,0 +1,267 @@
+package org.apache.flink.connector.dynamodb.table;
```
[the remark this answers is not in the document]
Comment (hlteoh37) [reply flink-connector-dynamodb/src/test/java/org/apache/flink/connector/dynamodb/table/DynamoDbDynamicSinkFactoryTest.java:1]: added

File: flink-connector-dynamodb/src/test/java/org/apache/flink/connector/dynamodb/table/DynamoDbDynamicSinkFactoryTest.java [hunk 21]
```diff
@@ -0,0 +1,267 @@
...
+                        DynamoDbDynamicSink.builder()
+                                .setTableName(DYNAMO_DB_TABLE_NAME)
+                                .setOverwriteByPartitionKeys(new HashSet<>(partitionKeys))
+                                .setDynamoDbClientProperties(dynamoDbClientProperties)
+                                .setPhysicalDataType(sinkSchema.toPhysicalRowDataType())
+                                .build();
+
+        assertThat(actualSink).usingRecursiveComparison().isEqualTo(expectedSink);
+        assertThat(actualSink.getChangelogMode(ChangelogMode.insertOnly()))
+                .isEqualTo(ChangelogMode.upsert());
```
[the remark this answers is not in the document]
Comment (hlteoh37) [reply flink-connector-dynamodb/src/test/java/org/apache/flink/connector/dynamodb/table/DynamoDbDynamicSinkFactoryTest.java:87]: The interface for `getChangelogMode()` says :
> When writing a dynamic table, the content can always be considered as a changelog (finite or infinite) for which all changes are written out continuously until the changelog is exhausted. The given ChangelogMode indicates the set of changes that the sink accepts during runtime.

I take this to mean that this method should return the mode that the sink supports. As such, we ignore the "requested" changelog mode given by the table planner.

**Review by vahmed-hamdy — APPROVED** [review]
LGTM, Thanks @hlteoh37

File: flink-sql-connector-dynamodb/pom.xml (hunk 18 above)
Comment (dannycranmer) [reply flink-sql-connector-dynamodb/pom.xml:89]: Ah, we had this issue for EFO too. I added a custom `execution.interceptors` for this https://github.com/apache/flink/commit/a330d4331b2af710b6531c55f427c4b14f037d78

**Review by dannycranmer — CHANGES_REQUESTED** [review]

File: flink-sql-connector-dynamodb/pom.xml (hunk 18 above)
Comment (hlteoh37) [reply flink-sql-connector-dynamodb/pom.xml:89]: Ok! done the same!

File: flink-sql-connector-dynamodb/pom.xml [hunk 22]
```diff
@@ -77,6 +77,7 @@ under the License.
 									<include>org.apache.httpcomponents:*</include>
 									<include>io.netty:*</include>
 									<include>commons-logging:commons-logging</include>
+									<include>commons-codec:commons-codec</include>
```
Comment (dannycranmer) [inline flink-sql-connector-dynamodb/pom.xml:80]: Looks like we will need to update the notice file for this

Comment (hlteoh37) [reply flink-sql-connector-dynamodb/pom.xml:80]: Good spot. Updated notice file

Used the following output from build

1.15.2
```
[INFO] --- maven-shade-plugin:3.2.4:shade (shade-flink) @ flink-sql-connector-dynamodb ---
[INFO] Including org.apache.flink:flink-connector-dynamodb:jar:1.0.0-SNAPSHOT in the shaded jar.
[INFO] Including org.apache.flink:flink-connector-base:jar:1.15.2 in the shaded jar.
[INFO] Including org.apache.flink:flink-connector-aws-base:jar:1.15.2 in the shaded jar.
[INFO] Including software.amazon.awssdk:netty-nio-client:jar:2.17.247 in the shaded jar.
[INFO] Including io.netty:netty-codec-http:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-codec-http2:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-codec:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-transport:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-resolver:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-common:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-buffer:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-handler:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-transport-classes-epoll:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-transport-native-unix-common:jar:4.1.77.Final in the shaded jar.
[INFO] Including org.reactivestreams:reactive-streams:jar:1.0.3 in the shaded jar.
[INFO] Including software.amazon.awssdk:sts:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:aws-query-protocol:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:dynamodb:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:aws-json-protocol:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:third-party-jackson-core:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:json-utils:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:protocol-core:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:profiles:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:sdk-core:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:auth:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:http-client-spi:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:regions:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:annotations:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:utils:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:aws-core:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:metrics-spi:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:apache-client:jar:2.17.247 in the shaded jar.
[INFO] Including org.apache.httpcomponents:httpclient:jar:4.5.13 in the shaded jar.
[INFO] Including commons-logging:commons-logging:jar:1.2 in the shaded jar.
[INFO] Including commons-codec:commons-codec:jar:1.11 in the shaded jar.
[INFO] Including org.apache.httpcomponents:httpcore:jar:4.4.13 in the shaded jar.
[INFO] Including software.amazon.awssdk:dynamodb-enhanced:jar:2.17.247 in the shaded jar.
[INFO] Excluding software.amazon.eventstream:eventstream:jar:1.0.1 from the shaded jar.
[INFO] Excluding org.apache.flink:flink-shaded-force-shading:jar:16.0 from the shaded jar.
[INFO] Excluding org.slf4j:slf4j-api:jar:1.7.36 from the shaded jar.
[INFO] No artifact matching filter io.netty:netty

```

1.16.0
```
[INFO] --- maven-shade-plugin:3.2.4:shade (shade-flink) @ flink-sql-connector-dynamodb ---
[INFO] Including org.apache.flink:flink-connector-dynamodb:jar:1.0.0-SNAPSHOT in the shaded jar.
[INFO] Including org.apache.flink:flink-connector-base:jar:1.16.0 in the shaded jar.
[INFO] Including org.apache.flink:flink-connector-aws-base:jar:1.16.0 in the shaded jar.
[INFO] Including software.amazon.awssdk:netty-nio-client:jar:2.17.247 in the shaded jar.
[INFO] Including io.netty:netty-codec-http:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-codec-http2:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-codec:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-transport:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-resolver:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-common:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-buffer:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-handler:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-transport-classes-epoll:jar:4.1.77.Final in the shaded jar.
[INFO] Including io.netty:netty-transport-native-unix-common:jar:4.1.77.Final in the shaded jar.
[INFO] Including org.reactivestreams:reactive-streams:jar:1.0.3 in the shaded jar.
[INFO] Including software.amazon.awssdk:sts:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:aws-query-protocol:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:dynamodb:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:aws-json-protocol:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:third-party-jackson-core:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:json-utils:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:protocol-core:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:profiles:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:sdk-core:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:auth:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:http-client-spi:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:regions:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:annotations:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:utils:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:aws-core:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:metrics-spi:jar:2.17.247 in the shaded jar.
[INFO] Including software.amazon.awssdk:apache-client:jar:2.17.247 in the shaded jar.
[INFO] Including org.apache.httpcomponents:httpclient:jar:4.5.13 in the shaded jar.
[INFO] Including commons-logging:commons-logging:jar:1.2 in the shaded jar.
[INFO] Including commons-codec:commons-codec:jar:1.11 in the shaded jar.
[INFO] Including org.apache.httpcomponents:httpcore:jar:4.4.13 in the shaded jar.
[INFO] Including software.amazon.awssdk:dynamodb-enhanced:jar:2.17.247 in the shaded jar.
[INFO] Excluding software.amazon.eventstream:eventstream:jar:1.0.1 from the shaded jar.
[INFO] Excluding org.apache.flink:flink-shaded-force-shading:jar:16.0 from the shaded jar.
[INFO] Excluding org.slf4j:slf4j-api:jar:1.7.36 from the shaded jar.
[INFO] No artifact matching filter io.netty:netty

```

**Review by dannycranmer — APPROVED** [review]
Thanks @hlteoh37, LGTM
